nav2_util/include/nav2_util/node_utils.hpp

```
@@ -81,47 +81,141 @@ rclcpp::Node::SharedPtr generate_internal_node(const std::string & prefix = "");
 /* Declares static ROS2 parameter with given type if it was not already declared.
  *
  * \param[in] node A node in which given parameter to be declared
+ * \param[in] parameter_name Name of the parameter
  * \param[in] param_type The type of parameter
- * \param[in] default_value Parameter value to initialize with
  * \param[in] parameter_descriptor Parameter descriptor (optional)
  */
 template<typename NodeT>
 void declare_parameter_if_not_declared(
   NodeT node,
-  const std::string & param_name,
+  const std::string & parameter_name,
   const rclcpp::ParameterType & param_type,
-  const rcl_interfaces::msg::ParameterDescriptor & parameter_descriptor =
-  rcl_interfaces::msg::ParameterDescriptor())
+  const ParameterDescriptor & parameter_descriptor = ParameterDescriptor())
+{
+  if (!node->has_parameter(parameter_name)) {
+    node->declare_parameter(parameter_name, param_type, parameter_descriptor);
+  }
+}
+
+/// Declares a parameter with the specified type if it was not already declared
+/// and returns its value if available.
+/* Declares a parameter with the specified type if it was not already declared.
+ * If the parameter was overridden, its value is returned, otherwise a
+ * parameter not set type is returned.
+ *
+ * \param[in] node A node in which given parameter to be declared
+ * \param[in] parameter_name Name of the parameter
+ * \param[in] param_type The type of parameter
+ * \param[in] parameter_descriptor Parameter descriptor (optional)
+ */
+template<typename NodeT>
+rclcpp::ParameterValue declare_or_get_parameter(
+  NodeT node,
+  const std::string & parameter_name,
+  const rclcpp::ParameterType & param_type,
+  const ParameterDescriptor & parameter_descriptor = ParameterDescriptor())
+{
+  if (node->has_parameter(parameter_name)) {
+    try {
+      return node->get_parameter(parameter_name).get_parameter_value();
+    } catch (const std::exception & exception) {
+      return rclcpp::ParameterValue();
```

A point of discussion: 

Right now the ROS 2 API is documented and expected to throw if you get a type-declared parameter but that parameter wasn't provided. In the places in Nav2 where we do type declarations we rely on the throw behavior to tell us a user didn't give a compulsory parameter to fail bringup. 

Personally, I think this is better to be like your other implementation below where we template by the parameter type to return rather than returning a `ParameterValue` or uninitialized one for a later throw to happen. If we're trying to wrap the details of parameter handling from application code, I think we should fully hide it and making the application code deal with a `rclcpp::ParameterValue` seems clunky.

In short, I think we should also template by the parameter value and have the returns return that direct value like the other implementation. Also, we should have this method throw when the parameter isn't provided rather than returning a default value.

---

I understand your concern, this format was used because the ROS declare parameter function doesn't throw when it's called with a type the first time, even if the parameter is not overridden, but returns a param not set.
As this function may also call get_parameter however, it is not possible to keep a 1-1 mapping in behavior; so if you prefer the exception, I can always raise an exception if the override isn't there.
I can also use a template and return the C++ type, however, in contrast with the latter case, the caller will have to specify the template type explicitly.